explore/src/main/scala/explore/config/AdvancedConfigurationPanel.scala

```diff
@@ -160,6 +172,94 @@ sealed abstract class AdvancedConfigurationPanelBuilder[
+      else
+        none
+  }
+
+  private def findMatrixDataFromRow(
+    basic:        S,
+    advanced:     T,
+    requirements: SpectroscopyRequirementsData,
+    row:          SpectroscopyModeRow
+  ): Option[ReadonlyData] = (basic, advanced, row.instrument) match {
+    case (ScienceModeBasic.GmosNorthLongSlit(bGrating, bFilter, bFpu),
+          ScienceModeAdvanced.GmosNorthLongSlit(aWavelength,
+                                                aGrating,
+                                                aFilter,
+                                                aFpu,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _
+          ),
+          GmosNorthSpectroscopyRow(rGrating, rFpu, rFilter)
+        ) =>
+      val wavelength = aWavelength.orElse(requirements.wavelength)
+      val grating    = aGrating.getOrElse(bGrating)
+      val filter     = aFilter.orElse(bFilter)
+      val fpu        = aFpu.getOrElse(bFpu)
+      if (grating === rGrating && filter === rFilter && fpu === rFpu)
+        ReadonlyData.build(row, wavelength)
+      else none
+    case (ScienceModeBasic.GmosSouthLongSlit(bGrating, bFilter, bFpu),
+          ScienceModeAdvanced.GmosSouthLongSlit(aWavelength,
+                                                aGrating,
+                                                aFilter,
+                                                aFpu,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _,
+                                                _
+          ),
+          GmosSouthSpectroscopyRow(rGrating, rFpu, rFilter)
+        ) =>
+      val wavelength = aWavelength.orElse(requirements.wavelength)
+      val grating    = aGrating.getOrElse(bGrating)
+      val filter     = aFilter.orElse(bFilter)
+      val fpu        = aFpu.getOrElse(bFpu)
+      if (grating === rGrating && filter === rFilter && fpu === rFpu)
+        ReadonlyData.build(row, wavelength)
+      else none
+    case _ => None
+  }
+
+  private def findMatrixData(
+    basic:        S,
+    advanced:     T,
+    requirements: SpectroscopyRequirementsData,
+    rows:         List[SpectroscopyModeRow]
+  ): Option[ReadonlyData] =
+    rows.flatMap(row => findMatrixDataFromRow(basic, advanced, requirements, row)).headOption
```

Comment:
I think we can avoid going through the whole list by using `collectFirst` (and building `ReadonlyData` afterwards).

Reply:
Good catch.